File: lib/api/models/chat_message.dart
// lib/models/chat_message.dart

class ChatMessage {
  final int? messageId;
  final int userId;
  final String content;
  final bool isAiResponse;
  final DateTime? createdAt;
  final int? relatedEntityId;
  final String? relatedEntityType;

  // Related user (optional)
  final Map<String, dynamic>? user;

  ChatMessage({
    this.messageId,
    required this.userId,
    required this.content,
    this.isAiResponse = false,
    this.createdAt,
    this.relatedEntityId,
    this.relatedEntityType,
    this.user,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      messageId: json['message_id'],
      userId: json['user_id'],
      content: json['content'],
      isAiResponse: json['is_ai_response'] ?? false,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      relatedEntityId: json['related_entity_id'],
      relatedEntityType: json['related_entity_type'],
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (messageId != null) 'message_id': messageId,
      'user_id': userId,
      'content': content,
      'is_ai_response': isAiResponse,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (relatedEntityId != null) 'related_entity_id': relatedEntityId,
      if (relatedEntityType != null) 'related_entity_type': relatedEntityType,
      if (user != null) 'user': user,
    };
  }

  ChatMessage copyWith({
    int? messageId,
    int? userId,
    String? content,
    bool? isAiResponse,
    DateTime? createdAt,
    int? relatedEntityId,
    String? relatedEntityType,
    Map<String, dynamic>? user,
  }) {
    return ChatMessage(
      messageId: messageId ?? this.messageId,
      userId: userId ?? this.userId,
      content: content ?? this.content,
      isAiResponse: isAiResponse ?? this.isAiResponse,
      createdAt: createdAt ?? this.createdAt,
      relatedEntityId: relatedEntityId ?? this.relatedEntityId,
      relatedEntityType: relatedEntityType ?? this.relatedEntityType,
      user: user ?? this.user,
    );
  }
}

File: final_project/models/chat_message.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey, Index
from sqlalchemy.sql import func
from final_project.db import Base

# SQLAlchemy model for storing chat messages between users and AI
class ChatMessageDB(Base):
    __tablename__ = "chat_message"

    message_id = Column(Integer, primary_key=True, autoincrement=True)  # Primary key
    user_id = Column(Integer, ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False)  # Link to UserDB
    content = Column(Text, nullable=False)  # Message content
    is_ai_response = Column(Boolean, default=False)  # Whether this message is from AI
    created_at = Column(DateTime, server_default=func.now())  # Message creation timestamp
    related_entity_id = Column(Integer)  # Related entity ID (optional)
    related_entity_type = Column(String(50))  # Related entity type (optional)

    __table_args__ = (
        Index("idx_chat_message_user_id", "user_id"),  # Index on user_id
        Index("idx_chat_message_created_at", "created_at"),  # Index on created_at
        Index("idx_chat_message_related_entity", "related_entity_id", "related_entity_type"),  # Composite index for related entities
    )

    # Relationships
    # user = relationship("UserDB", back_populates="chat_messages")
